tools/gfx/cuda/render-cuda.cpp

```diff
@@ -0,0 +1,1506 @@
+        return SLANG_OK;
+    }
+
+    virtual Result bindRootShaderObject(PipelineType pipelineType, ShaderObject* object) override
+    {
+        currentRootObject = dynamic_cast<CUDARootShaderObject*>(object);
+        if (currentRootObject)
+            return SLANG_OK;
+        return SLANG_E_INVALID_ARG;
+    }
+
+    virtual Result createProgram(const ShaderProgram::Desc& desc, ShaderProgram** outProgram) override
+    {
+        if (desc.kernelCount != 1)
+            return SLANG_E_INVALID_ARG;
+        RefPtr<CUDAShaderProgram> cudaProgram = new CUDAShaderProgram();
+        SLANG_CUDA_RETURN_ON_FAIL(cuModuleLoadData(&cudaProgram->cudaModule, desc.kernels[0].codeBegin));
+        SLANG_CUDA_RETURN_ON_FAIL(
+            cuModuleGetFunction(&cudaProgram->cudaKernel, cudaProgram->cudaModule, desc.kernels[0].entryPointName));
+        cudaProgram->kernelName = desc.kernels[0].entryPointName;
+        *outProgram = cudaProgram.detach();
+        return SLANG_OK;
+    }
+
+    virtual Result createComputePipelineState(const ComputePipelineStateDesc& desc, PipelineState** outState) override
+    {
+        RefPtr<CUDAPipelineState> state = new CUDAPipelineState();
+        state->shaderProgram = dynamic_cast<CUDAShaderProgram*>(desc.program);
+        *outState = state.detach();
+        return Result();
+    }
+
+    virtual void* map(BufferResource* buffer, MapFlavor flavor) override
+    {
+        return dynamic_cast<MemoryCUDAResource*>(buffer)->m_cudaMemory;
+    }
+
+    virtual void unmap(BufferResource* buffer) override
+    {
+        SLANG_UNUSED(buffer);
+    }
+
+    virtual void setPipelineState(PipelineType pipelineType, PipelineState* state) override
+    {
+        SLANG_ASSERT(pipelineType == PipelineType::Compute);
+        currentPipeline = dynamic_cast<CUDAPipelineState*>(state);
+    }
+
+    virtual void dispatchCompute(int x, int y, int z) override
+    {
+        // Find out thread group size from program reflection.
+        auto& kernelName = currentPipeline->shaderProgram->kernelName;
+        auto programLayout = dynamic_cast<CUDAProgramLayout*>(currentRootObject->layout.Ptr());
+        int kernelId = programLayout->getKernelIndex(kernelName.getUnownedSlice());
+        SLANG_ASSERT(kernelId != -1);
+        UInt threadGroupSize[3];
+        programLayout->getKernelThreadGroupSize(kernelId, threadGroupSize);
+
+        int sharedSizeInBytes;
+        cuFuncGetAttribute(
+            &sharedSizeInBytes,
+            CU_FUNC_ATTRIBUTE_SHARED_SIZE_BYTES,
+            currentPipeline->shaderProgram->cudaKernel);
+
+        // Copy global parameter data to the `SLANG_globalParams` symbol.
+        {
+            CUdeviceptr globalParamsSymbol = 0;
+            size_t globalParamsSymbolSize = 0;
+            cuModuleGetGlobal(
+                &globalParamsSymbol,
+                &globalParamsSymbolSize,
+                currentPipeline->shaderProgram->cudaModule,
+                "SLANG_globalParams");
```

Comment:
Note that a lot of the data here is stuff that should eventually be cached on the `shaderProgram` and/or `programLayout`.

E.g., there is no reason to call `cuModuleGetGlobal` on every launch, since the address of that symbol should not change for the lifetime of `shaderProgram`.